src/connection.js

```diff
@@ -605,6 +665,57 @@ class Connection extends EventEmitter {
       this.emit('charsetChange', token.newValue);
     });
 
+    this.tokenStreamParser.on('fedAuthInfo', (token) => {
+      const clientId = '7f98cb04-cd1e-40df-9140-3bf7e2cea4db';
+      if (token.fedAuthInfoData.stsurl && token.fedAuthInfoData.spn) {
+        this.fedAuthInfo.responsePending = true;
+        var context = new AuthenticationContext(token.fedAuthInfoData.stsurl);
+
+        const authentication = this.config.authentication;
+
+        context.acquireTokenWithUsernamePassword(token.fedAuthInfoData.spn, authentication.options.userName, authentication.options.password, clientId, (err, tokenResponse) => {
+          if (err) {
+            this.fedAuthInfo.responsePending = false;
+            this.loginError = ConnectionError('Security token could not be authenticated or authorized.', 'EFEDAUTH');
+          } else {
+            this.fedAuthInfo.responsePending = false;
+            this.fedAuthInfo.token = tokenResponse;
+          }
+        });
+      }
+    });
+
+    this.tokenStreamParser.on('featureExtAck', (token) => {
+      let fedAuthAck = undefined;
+      if (this.fedAuthInfo.fedAuthInfoRequested && (fedAuthAck = token.featureAckOpts.get(FEDAUTH_OPTIONS.FEATURE_ID)) === undefined) {
+        this.loginError = ConnectionError('Did not receive Active Directory authentication acknowledgement');
+        this.loggedIn = false;
+      }
+      if (fedAuthAck !== undefined) {
+        if (!this.fedAuthInfo.fedAuthInfoRequested) {
+          this.loginError = ConnectionError('Did not request Active Directory authentication, but received the acknowledgment');
+          this.loggedIn = false;
+        }
+        switch (this.fedAuthInfo.fedAuthLibrary) {
+          case FEDAUTH_OPTIONS.LIBRARY_ADAL:
+            if (0 !== fedAuthAck.length) {
+              this.loginError = ConnectionError(`Active Directory authentication acknowledgment for ${this.fedAuthInfo.method} authentication method includes extra data`);
+              this.loggedIn = false;
+            }
+            break;
+          default:
+            this.loginError = ConnectionError('Attempting to use unknown Active Directory authentication library');
+            this.loggedIn = false;
+        }
+      }
+      else {
+        this.loginError = ConnectionError('Received acknowledgement for unknown feature');
+        this.loggedIn = false;
+      }
+
```

Comment:
If `fedAuthAck = undefined` and `this.fedAuthInfo.fedAuthInfoRequested = true`, then there are 2 conditions that are met, one with error message 'Did not receive Active Directory authentication acknowledgement' and other with 'Received acknowledgement for unknown feature'.

Can you restructure all these error messages such that one will not overwrite the other?